File: src/main/java/com/heekng/celloct/entity/WorkTime.java
package com.heekng.celloct.entity;

import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import javax.persistence.Column;
import javax.persistence.Embeddable;
import java.time.Duration;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.Period;

@Embeddable
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class WorkTime {

    @Column(nullable = false)
    private LocalDate workDate;
    @Column(nullable = false)
    private LocalDateTime startDate;
    @Column(nullable = false)
    private LocalDateTime endDate;

    @Builder
    public WorkTime(LocalDate workDate, LocalDateTime startDate, LocalDateTime endDate) {
        this.workDate = workDate;
        this.startDate = startDate;
        this.endDate = endDate;
    }

    public void changeWorkTime(LocalDateTime startDate, LocalDateTime endDate) {
        this.startDate = startDate;
        this.endDate = endDate;
    }

    public long workTimeHour() {
        Duration duration = Duration.between(startDate, endDate);
        long seconds = duration.getSeconds();
        return seconds / 3600;
    }

    public long workTimeMinute() {
        Duration duration = Duration.between(startDate, endDate);
        long seconds = duration.getSeconds();
        return seconds % 3600;
    }

}

File: src/main/kotlin/com/heekng/celloct/entity/WorkTime.kt
package com.heekng.celloct.entity

import java.time.Duration
import java.time.LocalDate
import java.time.LocalDateTime
import javax.persistence.Column
import javax.persistence.Embeddable

@Embeddable
class WorkTime(
    @Column(nullable = false, name = "work_date")
    val workDate: LocalDate,

    @Column(nullable = false, name = "start_date")
    var startDate: LocalDateTime,

    @Column(nullable = false, name = "end_date")
    var endDate: LocalDateTime,
) {

    companion object {
        fun fixture(
            workDate: LocalDate,
            startDate: LocalDateTime,
            endDate: LocalDateTime,
        ): WorkTime {
            return WorkTime(
                workDate = workDate,
                startDate = startDate,
                endDate = endDate,
            )
        }
    }

    // TODO: 2022/08/22 값타입 내에서 값을 변경하는 로직이 포함됨. 이부분 수정이 필요함
    fun changeWorkTime(
        startDate: LocalDateTime,
        endDate: LocalDateTime,
    ) {
        this.startDate = startDate
        this.endDate = endDate
    }

    fun workTimeHour(): Long {
        val seconds = Duration.between(startDate, endDate).seconds
        return seconds / 3600
    }

    fun workTimeMinute(): Long {
        val seconds = Duration.between(startDate, endDate).seconds
        return seconds % 3600 / 60
    }
}